src/odbc/unittests/tvp.c

```diff
@@ -0,0 +1,188 @@
+
+    // Population of the StrLen_or_IndPtr buffer can be deferred
+    numRows = MAX_ROWS;
+
+	CHKExecDirect(T("{CALL TestTVPProc(?)}"), SQL_NTS, "S");
+
+    // Ensure that we successfully add 5 rows
+	odbc_command("SELECT COUNT(*) FROM TVPTable");
+
+	CHKFetch("SI");
+
+    CHKGetData(1, SQL_C_CHAR, outputBuffer, sizeof(outputBuffer), &lenBuffer, "S");
+    if(strcmp((char *) outputBuffer, "5") != 0) {
+        fprintf(stderr, "Wrong number of columns inserted, expected %ld, got %s\n", numRows, outputBuffer);
+        exit(1);
+    }
+
+	CHKFetch("No");
+
+    odbc_command("DROP PROC TestTVPProc");
+    odbc_command("DROP TYPE TVPType");
+    odbc_command("DROP TABLE TVPTable");
+}
+
+/*
+ * Test TVP usage with more parameter types, using a TVP of 2 columns and 5 rows
+ */
+static void
+TestTVPInsert2(void)
+{
+    SQLCHAR * tableName[MAX_STRING_LENGTH]; // Test explicit schema declaration
+    SQLLEN numRows;
+
+    strncpy((char *) tableName, "dbo.TVPType2", MAX_STRING_LENGTH);
+
+	odbc_command("IF OBJECT_ID('TestTVPProc2') IS NOT NULL DROP PROC TestTVPProc2");
+	odbc_command("DROP TYPE IF EXISTS TVPType2");
+	odbc_command("DROP TABLE IF EXISTS TVPTable2");
+
+	odbc_command("CREATE TABLE TVPTable2 (Num NUMERIC(10, 5), Bin BINARY(10))");
+	odbc_command("CREATE TYPE TVPType2 "
+        "AS TABLE (vNum NUMERIC(10, 5), vBin VARBINARY(10))");
+	odbc_command("CREATE PROCEDURE TestTVPProc2 (@TVPParam TVPType2 READONLY) "
+        "AS INSERT INTO TVPTable2 SELECT vNum, vBin FROM @TVPParam");
+
+	CHKBindParameter(1, SQL_PARAM_INPUT, SQL_C_DEFAULT, SQL_SS_TABLE, MAX_ROWS, 0, tableName, SQL_NTS, &numRows, "S");
+
+    CHKSetStmtAttr(SQL_SOPT_SS_PARAM_FOCUS, (SQLPOINTER) 1, SQL_IS_INTEGER, "S");
+
+	CHKBindParameter(1, SQL_PARAM_INPUT, SQL_C_NUMERIC, SQL_NUMERIC,
+            10, 5, numericCol, sizeof(SQL_NUMERIC_STRUCT), lNumericCol, "S");
+	CHKBindParameter(2, SQL_PARAM_INPUT, SQL_C_BINARY, SQL_BINARY,
+            MAX_STRING_LENGTH, 4, binCol, MAX_STRING_LENGTH, lBinCol, "S");
+
+    CHKSetStmtAttr(SQL_SOPT_SS_PARAM_FOCUS, (SQLPOINTER) 0, SQL_IS_INTEGER, "S");
+
+    // Population of the StrLen_or_IndPtr buffer can be deferred
+    numRows = MAX_ROWS;
+
+	CHKExecDirect(T("{CALL TestTVPProc2(?)}"), SQL_NTS, "S");
+
+    // Ensure that we successfully add 5 rows
+	odbc_command("SELECT COUNT(*) FROM TVPTable2");
+
+	CHKFetch("SI");
+
+    CHKGetData(1, SQL_C_CHAR, outputBuffer, sizeof(outputBuffer), &lenBuffer, "S");
+    if(strcmp((char *) outputBuffer, "5") != 0) {
+        fprintf(stderr, "Wrong number of columns inserted, expected %ld, got %s\n", numRows, outputBuffer);
+        exit(1);
+    }
+
+	CHKFetch("No");
+
+    odbc_command("DROP PROC TestTVPProc2");
+    odbc_command("DROP TYPE TVPType2");
+    odbc_command("DROP TABLE TVPTable2");
+}
+
+int
+main(int argc, char *argv[])
+{
+	odbc_connect();
+
+    setup();
+
+    TestTVPInsert();
+    TestTVPInsert2();
+
+	CHKCloseCursor("SI");
+    odbc_disconnect();
+
+	return 0;
+}
```

Comment:
Seems that their new file convention is to have extra ending new line. Probably good to follow that.